### ISSUE TITLE
Invalid hint for cascade

### ISSUE THREAD

**Issue by zoechi** [dart-lang/linter#339]
```dart
    final resource = (entry.resource as Practitioner);
    resource.identifier = [
      new Identifier()
        ..system = 'urn:system'
        ..value = 'mdc_${resource.id}'
    ];
```
With cascades `'mdc_${resource.id}'` doesn't work

> Dart VM version: 1.21.0-edge.640b5ca80a2739d2db6466834778c60147ad15c0 (Mon Nov 21 12:35:27 2016) on "linux_x64"

**Comment by zoechi** [comment]
It's also not clear to me where I can add

```dart
// ignore: cascade_invocations  
```

to silence this hint. 

The quickfix adds it before the 2nd line but that doesn't silence the hint.

**Comment by alexeieleusis** [comment]
The comment needs to be one line above the lint you want to silence.

I just tried to reproduce this with the following snippet but I didn't lint:

```
class ClassWithStrings {
  String first;
  String second;
  Foo foo;

  void someThing() {
    final variable = new ClassWithStrings()
        ..first = 'constant'
        ..second = '${foo.bar}';
  }
}
```

Do you have a self contained example? I don't know the classes you are using and might be related with the failure.

**Comment by zoechi** [comment]
```dart
class Identifier {
  String value;
  String system;
}

class Resource {
  String id;
  List<Identifier> identifier;
}

class Practitioner extends Resource{
}

class Entry {
  Resource resource;
}

void main() {
  final entry = new Entry(); 
  final resource = (entry.resource as Practitioner);
  // ignore: cascade_invocations
  resource.identifier = [
    new Identifier()
      ..system = 'urn:system'
      ..value = 'mdc_${resource.id}'
  ];
}
```
I stil get the linter warning for `resource.identifier = [...`
```
analyzer:
  strong-mode:
    implicit-casts: false
    implicit-dynamic: false

#  language:
#    enableGenericMethods: true

  exclude:
#    - 'build/**'
#    - '**.g.dart'
    - 'bazel-*'

linter:
  rules:
# Error Rules
    - avoid_empty_else
    - cancel_subscriptions
    - close_sinks
    - comment_references
    - control_flow_in_finally
    - empty_statements
    - hash_and_equals
    - invariant_booleans
    - iterable_contains_unrelated_type
    - list_remove_unrelated_type
    - literal_only_boolean_expressions
    - test_types_in_equals
    - throw_in_finally
    - unrelated_type_equality_checks
    - valid_regexps

# Style Rules
    - always_declare_return_types
#    - always_specify_types
    - annotate_overrides
#    - avoid_as
    - avoid_init_to_null
    - avoid_return_types_on_setters
    - await_only_futures
    - camel_case_types
    - cascade_invocations
    - constant_identifier_names
    - empty_catches
    - empty_constructor_bodies
    - implementation_imports
    - library_names
    - library_prefixes
    - non_constant_identifier_names
    - one_member_abstracts
    - only_throw_errors
    - overridden_fields
    - package_api_docs
    - package_prefixed_library_names
    - prefer_is_not_empty
    - parameter_assignments
    - prefer_final_fields
    - prefer_final_locals
#    - public_member_api_docs
    - slash_for_doc_comments
#    - sort_constructor_first
    - sort_unnamed_constructors_first
    - super_goes_last
    - type_annotate_public_apis
    - type_init_formals
    - unawaited_futures
    - unnecessary_brace_in_string_interp
    - unnecessary_getters_setters

# Pub Rules
    - package_names
```

**Comment by alexeieleusis** [comment]
Sorry, I didn't understand where the lint was, you need resource to be declared in a previous statement so string interpolation works, I changed it to this, I don't like it but it does not lint:

```
void main() {
  final entry = new Entry();
  Practitioner resource;
  resource = (entry.resource as Practitioner)
      ..identifier = [
    new Identifier()
      ..system = 'urn:system'
      ..value = 'mdc_${resource.id}'
  ];
}
```

**Comment by bwilkerson** [comment]
The bug here is that we shouldn't be producing a lint error at all if the variable is referenced in the code that would be cascaded (because converting it to a cascade would change the semantics). It's a false positive and we need to fix it so that we don't have false positives.
